kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Marshall Wace <opensource@mwam.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.mwam.kafkakewl.metrics.services
+
+import com.mwam.kafkakewl.metrics.MainConfig
+import com.mwam.kafkakewl.metrics.domain.{KafkaTopicPartition, KafkaTopicPartitionInfo, KafkaTopicPartitionInfoChanges}
+import zio.kafka.consumer.Consumer
+import zio.{Duration, Fiber, Schedule, Task, UIO, URIO, ZIO, ZLayer}
+import zio.stream.{ZSink, ZStream}
+import zio.metrics.*
+import zio.metrics.Metric.Counter
+
+import scala.collection.immutable.Map
+
+object Helper {
+  def newTopicMetric(topic: String): Metric.Counter[Long] = Metric.counter("kafkakewl_vnext_topic_high_offset").tagged("topic", topic)
+  def newPartitionMetric(topic: String, partition: Int): Metric.Counter[Long] =
+    Metric.counter("kafkakewl_vnext_topic_partition_high_offset").tagged("topic", topic).tagged("partition", partition.toString)
+}
+
+case class MetricMap(metricMap: Map[String, (Map[Int, (Metric.Counter[Long], Long)], Metric.Counter[Long])]) {
+
+  def updateWithTopicParitionChanges(
+      changes: KafkaTopicPartitionInfoChanges,
+      config: MetricExposerConfig
+  ): (MetricMap, List[(Metric.Counter[Long], Long)]) = {
+    val addedOrUpdated = changes.addedOrUpdated
+    val removed = changes.removed
+    var newMetricMap = metricMap
+    var metricUpdates: List[(Metric.Counter[Long], Long)] = List.empty
+    // Add new metrics and update stale ones
+    for ((topicPartition, partitionInfo) <- addedOrUpdated) {
+
+      newMetricMap = newMetricMap.updatedWith(topicPartition.topic) {
+        case Some((topicMap: Map[Int, (Metric.Counter[Long], Long)], aggregateCounter: Metric.Counter[Long])) =>
+          val newTopicMap = topicMap.updatedWith(topicPartition.partition) {
+            case Some((counter, lastValue)) =>
+              val increase = partitionInfo.endOffset - lastValue
+              metricUpdates = metricUpdates :+ (aggregateCounter, increase)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, increase)
+              }
+              Some((counter, partitionInfo.endOffset))
+            case None =>
+              val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+              metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
+              if (config.perPartitionMetrics) {
+                metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
+              }
+              Some((counter, partitionInfo.endOffset))
+          }
+          Some((newTopicMap, aggregateCounter))
+        case None =>
+          val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
+          val newTopicMap = Map.newBuilder
```

Comment:
This can be simplified to `Map(topicPartition.partition ->(counter, partitionInfo.endOffset))`

Reply:
No longer applicable given changes